src/compiler/crystal/syntax/lexer.cr

```diff
@@ -1456,567 +1450,191 @@ module Crystal
-      next_char
 
-      num = 0_u64
-      num_size = first_digit = 0
-      while true
-        char = next_char
-        if '0' <= char <= '7'
-          num = (num << 3) | (char - '0')
-          if num_size == 0
-            first_digit = num
-            num_size += 1 if char != '0'
-          else
-            num_size += 1
-          end
-        elsif char == '_'
-        else
-          break
+        # Skip prefix (b, o, x)
+        unless base == 10
+          next_char
+          pos_after_prefix = current_pos
+          # Enforce number after prefix (disallow ex. "0x", "0x_1")
+          raise("numeric literal without digits", @token, (current_pos - start)) unless String::CHAR_TO_DIGIT[current_char.ord].to_u8! < base
         end
       end
 
-      # 0o177777_77777777_77777777 is the largest UInt64.
-      num = nil if {num_size, first_digit} > {22, 0o1}
-      finish_scan_prefixed_number num, negative, start
-    end
-
-    def scan_hex_number(start, negative = false)
-      next_char
-
-      num = 0_u64
-      num_size = 0
-      while true
-        char = next_char
-        if char == '_'
-        else
-          hex_value = char_to_hex(char) { nil }
-          if hex_value
-            num = (num << 4) | hex_value
-            if num_size > 0 || char != '0'
-              num_size += 1
-            end
-          else
-            break
-          end
+      # Consume number
+      loop do
+        while String::CHAR_TO_DIGIT[current_char.ord].to_u8! < base
+          number_size += 1 unless number_size == 0 && current_char == '0'
+          next_char
+          last_is_underscore = false
         end
-      end
-
-      # 0xFFFF_FFFF_FFFF_FFFF is the longest UInt64.
-      num = nil if num_size > 16
-      finish_scan_prefixed_number num, negative, start
-    end
 
-    def finish_scan_prefixed_number(num : Int?, negative : Bool, start : Int32)
-      if num.nil? # Doesn't even fit in UInt64
         case current_char
-        when 'i'
-          consume_int_suffix
-        when 'u'
-          consume_uint_suffix
-        else
-          @token.number_kind = :u64
-        end
-        case @token.number_kind
-        when :i8, :i16, :i32, :i64, :i128
-          type_name = "Int" + @token.number_kind.to_s[1..]
-        when :u8, :u16, :u32, :u64, :u128
-          type_name = "UInt" + @token.number_kind.to_s[1..]
+        when '_'
+          raise("consecutive underscores in numbers aren't allowed", @token, (current_pos - start)) if last_is_underscore
+          has_underscores = last_is_underscore = true
+        when '.'
+          raise("trailing '_' in number", @token, (current_pos - start)) if last_is_underscore
+          break if is_decimal || base != 10 || !peek_next_char.in?('0'..'9')
+          is_decimal = true
+        when 'e', 'E'
+          last_is_underscore = false
+          break if is_e_notation || base != 10
+          is_e_notation = is_decimal = true
+          next_char if peek_next_char.in?({'+', '-'})
```

Comment:
Style: just `.in?('+', '-')` works - without the explicit tuple. There are many occurrences of this in this PR, please change them all.

Reply:
Thanks, I didn't know that!